packages/block-library/src/block/edit.js

```diff
@@ -6,207 +6,158 @@ import { partial } from 'lodash';
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
-import { Placeholder, Spinner, Disabled } from '@wordpress/components';
-import { withSelect, withDispatch } from '@wordpress/data';
-import { __ } from '@wordpress/i18n';
 import {
 	BlockEditorProvider,
 	BlockList,
 	WritingFlow,
 } from '@wordpress/block-editor';
-import { compose } from '@wordpress/compose';
 import { parse, serialize } from '@wordpress/blocks';
+import { Placeholder, Spinner, Disabled } from '@wordpress/components';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
 import ReusableBlockEditPanel from './edit-panel';
 
-class ReusableBlockEdit extends Component {
-	constructor( { reusableBlock } ) {
-		super( ...arguments );
-
-		this.startEditing = this.startEditing.bind( this );
-		this.stopEditing = this.stopEditing.bind( this );
-		this.setBlocks = this.setBlocks.bind( this );
-		this.setTitle = this.setTitle.bind( this );
-		this.save = this.save.bind( this );
-
-		if ( reusableBlock ) {
-			// Start in edit mode when we're working with a newly created reusable block
-			this.state = {
-				isEditing: reusableBlock.isTemporary,
-				title: reusableBlock.title,
-				blocks: parse( reusableBlock.content ),
-			};
-		} else {
-			// Start in preview mode when we're working with an existing reusable block
-			this.state = {
-				isEditing: false,
-				title: null,
-				blocks: [],
+export default function ReusableBlockEdit( { attributes, isSelected } ) {
+	const { ref } = attributes;
+
+	const {
+		reusableBlock,
+		isFetching,
+		isSaving,
+		isTemporary,
+		blocks,
+		canUpdateBlock,
+		settings,
+		title,
+	} = useSelect(
+		( select ) => {
+			const { canUser } = select( 'core' );
+			const { getSettings } = select( 'core/block-editor' );
+			const {
+				__experimentalGetReusableBlock: getReusableBlock,
+				__experimentalIsFetchingReusableBlock: isFetchingReusableBlock,
+				__experimentalIsSavingReusableBlock: isSavingReusableBlock,
+			} = select( 'core/editor' );
+			const _reusableBlock = getReusableBlock( ref );
+
+			return {
+				reusableBlock: _reusableBlock,
+				isFetching: isFetchingReusableBlock( ref ),
+				isSaving: isSavingReusableBlock( ref ),
+				isTemporary: _reusableBlock?.isTemporary ?? null,
+				blocks: _reusableBlock ? parse( _reusableBlock.content ) : null,
```

Comment:
`parse` here is concerning, since the `useSelect` selector will be called very frequently for any reusable block which is mounted in the page (every state change for every store). The selector should be as efficient as possible, and `parse` isn't an especially trivial operation. In the previous implementation, we apparently relied on memoization of `__experimentalGetParsedReusableBlock` to avoid calling `parse` more than necessary. Is there any reason that was dropped here?

Reply:
It was an oversight on my part. Thanks for catching! I've restored the usage of `__experimentalGetParsedReusableBlock`.

Reply:
It turns out that trying to use `__experimentalGetParsedReusableBlock( _reusableBlock.id )` instead of `parse( _reusableBlock.content )` was causing end-to-end test failures and broken behavior in my testing. I'm not sure why, but for now, I've reverted back to using the `parse` approach.

Reply:
Ok, but that brings us back to the original concern in https://github.com/WordPress/gutenberg/pull/21427#discussion_r424666982 . 

Furthermore, we shouldn't be okay to accept a solution because we don't understand a failing of the preferred option. And `__experimentalGetParsedReusableBlock` must exist for a reason, or at least: We should want to use it or eliminate it consistently.

What errors were you seeing?

Reply:
I only reverted back to `parse` temporarily to focus on the [other issue](https://github.com/WordPress/gutenberg/pull/21427#discussion_r424667843) you brought up. I don't think we should merge this until we figure out this problem, too.

I'll push a commit redoing the switch back to `__experimentalGetParsedReusableBlock` to make it easier to test once I'm finished with the other issue.

(Testing stuff has taken a really long time ever since [`wp-env` stopped working on Linux](https://github.com/WordPress/gutenberg/issues/20180) and I've had to rely on `npm run package-plugin` or gutenberg.run.)

Reply:
Ah, fair enough, sorry for jumping to conclusions 😄

Reply:
Alright, I've redone the commit to switch to `__experimentalGetParsedReusableBlock`. It seems to completely break reusable block creation. I have no idea why. But anyway, it should be easy for others to test and help figure it out now.